[cm11a] Initial contribution and 2nd code review

BobRak opened the pull request:
This should be considered an extension of pull #1724. This fork is the result of code changes identified in the PR.

The cm11a is a serial computer interface for X-10 devices. The goal of this request is to have this binding merged into openHAB. Or, at least get constructive feedback on changes to be made. It has been running on my openHAB system for several months. It has also been available in my [repo](https://github.com/BobRak/OpenHAB-cm11a)and tested by other users.

martinvw commented on addons/binding/org.openhab.binding.cm11a/OSGI-INF/CM11AHandlerFactory.xml:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2014-2016 by the respective copyright holders.
```
It seems that this file has a missing, outdated or otherwise non standard copyright header, can you fix all of them by running 

`mvn license:format` 

Please commit only the ones for your binding.

martinvw commented on addons/binding/org.openhab.binding.cm11a/src/etc/header.txt:
```diff
@@ -0,0 +1,6 @@
+Copyright (c) 2014-${year} by the respective copyright holders.
```
Imho this can be removed / is not needed

martinvw commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/CM11ABindingConstants.java:
```diff
@@ -0,0 +1,44 @@
+ * @author bob raker - Initial contribution
+ */
+/**
+ * @author Bob
```
Duplicate class level docblock

martinvw commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aAbstractHandler.java:
```diff
@@ -0,0 +1,185 @@
+        int newPercent = curPercent - dimsPercent;
+        newPercent = Math.max(newPercent, 0);
+        newPercent = Math.min(newPercent, 100);
+        currentState = new PercentType(newPercent);
```
Unneeded local before return

BobRak replied:
Are you referring to currentState? This is a class variable. Maybe I am misunderstanding what the problem is, sorry.

martinvw replied:
Nope, I was wrong :-)

martinvw commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aAbstractHandler.java:
```diff
@@ -0,0 +1,185 @@
+     * @param thing The "Thing" to be handled
+     */
+    public Cm11aAbstractHandler(Thing thing) {
+        super(thing);
```
Please assign the thing here and make the field private

BobRak replied:
I removed the class variable "thing" because it exists in BaseThingHandler.

martinvw commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aApplianceHandler.java:
```diff
@@ -0,0 +1,95 @@
+
+        Cm11aBridgeHandler cm11aHandler = (Cm11aBridgeHandler) bridge.getHandler();
+        if (cm11aHandler != null && cm11aHandler.getThing().getStatus().equals(ThingStatus.ONLINE)) {
+            if (command.equals(OnOffType.ON)) {
```
Either compare with == (its an enum) or put the constant at the left.

martinvw commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/internal/X10ReceivedData.java:
```diff
@@ -0,0 +1,153 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+/**
```
Remove this dangling doc block.

martinvw commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aApplianceHandler.java:
```diff
@@ -0,0 +1,95 @@
+                X10Interface x10Interface = cm11aHandler.getX10Interface();
+                x10Interface.scheduleHWUpdate(this);
+            } else {
+                logger.debug("Received invalid command for switch {} command: {}", houseUnitCode, command.toString());
```
You don't have to call toString it will be called automatically, but only when the logging-level actually applies.

Please check all

martinvw commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aApplianceHandler.java:
```diff
@@ -0,0 +1,95 @@
+                    updateState(channelUID, desiredState);
+                    setCurrentState(desiredState);
+                }
+            } catch (Exception e) {
```
You used to catch two exceptions I suggested to replace it like follow any reason you changed it to this:

```java
} catch (InvalidAddressException | IOException e) {
    logger.error( "cm11a was not able to update the cm11a because of an exception. Check your hardware.", e);
}
```

Please also remove the `throws IOException, InvalidAddressException` and from the interface, please also validate that you do not catch them on places where they will not occur.

martinvw commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aBridgeHandler.java:
```diff
@@ -0,0 +1,229 @@
+ * for a description of OpenHAB bridges. This gets called first and is responsible for setting up the handler. Mostly it
+ * loads the 10Interface class which does all of the heavy lifting.
+ *
+ * @author bob raker - Initial contribution
```
Can you do a quick search replace to spell your own name with capitals :-)

martinvw commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aBridgeHandler.java:
```diff
@@ -0,0 +1,229 @@
+        synchronized (rd) {
+            for (Thing thing : things) {
+                String houseUnitCode = (String) thing.getConfiguration().get("HouseUnitCode");
+                for (String huc : rd.getAddr()) {
```
Maybe give this a more clear name like `messageHouseUnitCode`, `messageAddress`, `targetAddress`

martinvw commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aLampHandler.java:
```diff
@@ -0,0 +1,152 @@
+                    // Hardware update failed, log
+                    logger.error("cm11a failed to update device: {}", houseUnitCode);
+                }
+            } catch (Exception e) {
```
Same remark about the caught exceptions please check all.

martinvw commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/internal/X10Interface.java:
```diff
@@ -0,0 +1,788 @@
+                    logger.info("Serial receive timeout not supported by this driver.");
+                }
+            } catch (PortInUseException e) {
+                logger.warn("Serial port {} is in use by another application ({})", portId.getName(), e.currentOwner);
```
I did a quick check, but do you properly update the bridge connection state based on this exceptions. Please validate this and also make the bridge status 'always' correctly represents the real status.

martinvw commented:
Can you update your commit message to include a `Signed-off-by` (to certify that you wrote it or otherwise have the right to pass it on as an open-source), see http://docs.openhab.org/developers/contributing/contributing.html#sign-your-work

https://stackoverflow.com/a/8981216/3656774

BobRak commented:
I think I have now addressed all of your comments.

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/README.md:
```diff
@@ -0,0 +1,101 @@
+    Thing switch SwitchA1 [ HouseUnitCode="A1" ]
+    Thing dimmer DimmerA2 [ HouseUnitCode="A2" ]
+
+### Example
```
Instead of doing separate examples for things/items/sitemaps I would prefer a single "Full example" as we have it for all the other bindings. No need to explain the syntax in detail as this should be common to all bindings (and thus explained once at a central place in our docs).

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/README.md:
```diff
@@ -0,0 +1,101 @@
+    Switch SwitchA1  "Kitchen Plug"   <light>  (someGroup)  { channel="cm11a:switch:MyCm11a:SwitchA1:switchstatus" }
+    Dimmer DimmerA2  "Porch lights"   <slider> (someGroup)  { channel="cm11a:dimmer:MyCm11a:DimmerA2:lightlevel" }
+
+## Hardware - cm11a
```
Wouldn't this section be best as an introduction at the top?

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/README.md:
```diff
@@ -0,0 +1,101 @@
+
+### Use of serial port
+
+The binding opens the serial port when it starts and keeps it open until the binding is terminated. If the serial port is disconnected a reconnect will be attempted the next time it is needed.  Therefore, other applications should not attempt to use the port when OpneHAB is running. However, another program could load macros into the cm11a before OpenHAB starts. 
```
please correct "OpenHAB" to be "openHAB".

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/README.md:
```diff
@@ -0,0 +1,101 @@
+
+### cm11a macros
+
+The cm11a is also able to store a schedule and control modules based on that schedule. That functionality in not currently supported by or used by this binding. This binding doesn't clear macros from the cm11a so other programs could load macros before OpenHAB is started. If you want to do scheduling using OpenHAB you should be sure there are no macros in the cm11a. The `heyu clear` command can b ...
```
please correct "OpenHAB" to be "openHAB".

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/README.md:
```diff
@@ -0,0 +1,101 @@
+
+## Known issues
+
+1. When OpenHAB starts up it doesn't restore the last state of each module. And, the cm11a does not provide a discovery service. Therefore it assumes everything off.
```
please correct "OpenHAB" to be "openHAB".

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/README.md:
```diff
@@ -0,0 +1,101 @@
+
+1. [CM11A (X10) Protocol Document](http://wanderingsamurai.net/electronics/cm11a-x10-protocol-document)
+2. [Heyu - control software for the cm11a](http://www.heyu.org/)
+3. cm11a Controllers are available for purchase from several sites on the internet including EBay and Amazon.
```
remove " including EBay and Amazon."

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/.classpath:
```diff
@@ -0,0 +1,11 @@
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="src" path="src/main/java">
+		<attributes>
+			<attribute name="org.eclipse.jdt.launching.CLASSPATH_ATTR_LIBRARY_PATH_ENTRY" value="org.openhab.binding.cm11a/lib"/>
```
what is this about? I do not see any /lib folder in the project, so I guess that can be removed.

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,89 @@
+                <label>X10 House and Unit code</label>
+                <description>Address of dimmer in the X10 system</description>
+            </parameter>
+            <parameter name="Location" type="text">
```
This parameter is nowhere mentioned in the documentation and it sounds to me as if it does not really make sense as a configuration parameter. Can it be removed?

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,89 @@
+                <label>X10 House and Unit code</label>
+                <description>Address of switch in the X10 system</description>
+            </parameter>
+            <parameter name="Location" type="text">
```
This parameter is nowhere mentioned in the documentation and it sounds to me as if it does not really make sense as a configuration parameter. Can it be removed?

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/pom.xml:
```diff
@@ -0,0 +1,19 @@
+
+  <groupId>org.openhab.binding</groupId>
+  <artifactId>org.openhab.binding.cm11a</artifactId>
+  <version>2.2.0-SNAPSHOT</version>
```
remove the version here

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/README.md:
```diff
@@ -0,0 +1,101 @@
+
+The binding currently supports the following thing types:
+
+* switch
```
Especially, you should mention here, what channels they offer.

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aAbstractHandler.java:
```diff
@@ -0,0 +1,180 @@
+ * This is an abstract base class for the "Thing" handlers (i.e. Cm11aApplianceHandler and Cm11aLampHandler).
+ * It is not used the the Bridge handler (Cm11aHandler)
+ *
+ * @author Bob Raker
```
add "- Initial contribution"

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/config/Cm11aConfig.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Configuration constants for Cm11A
+ *
+ * @author Bob Raker
```
add "- Initial contribution"

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aAbstractHandler.java:
```diff
@@ -0,0 +1,180 @@
+
+/**
+ * This is an abstract base class for the "Thing" handlers (i.e. Cm11aApplianceHandler and Cm11aLampHandler).
+ * It is not used the the Bridge handler (Cm11aHandler)
```
typo: "the the" -> "by the"

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aAbstractHandler.java:
```diff
@@ -0,0 +1,180 @@
+     */
+    static final int X10_DIM_INCREMENTS = 22;
+
+    private Logger logger = LoggerFactory.getLogger(Cm11aAbstractHandler.class);
```
add final

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aAbstractHandler.java:
```diff
@@ -0,0 +1,180 @@
+            return;
+        }
+
+        houseUnitCode = (String) config.get("HouseUnitCode");
```
Use constants instead of strings

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aAbstractHandler.java:
```diff
@@ -0,0 +1,180 @@
+
+    @Override
+    public void dispose() {
+        houseUnitCode = "";
```
shouldn't it be rather set to null?

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aAbstractHandler.java:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    @Override
+    public abstract void handleCommand(ChannelUID channelUID, Command command);
```
why overwrite it in an abstract class?

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aAbstractHandler.java:
```diff
@@ -0,0 +1,180 @@
+     * Add the specified number of X10 "dims" from the current state
+     *
+     * @param dims The number of dims to remove
+     * @return
```
return what?

BobRak replied:
Nothing is returned, remove comment

kaikreuzer replied:
!? But the method has a return statement...?

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aApplianceHandler.java:
```diff
@@ -0,0 +1,89 @@
+/**
+ * Handler for Appliance (also called Switch) modules. These modules only support ON and OFF states
+ *
+ * @author Bob Raker
```
add "- Initial contribution"

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/Cm11aLampHandler.java:
```diff
@@ -0,0 +1,145 @@
+/**
+ * Handler for Lamp modules. These modules support ON, OFF and brightness level states
+ *
+ * @author Bob Raker
```
add "- Initial contribution"

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/handler/ReceivedDataListener.java:
```diff
@@ -0,0 +1,29 @@
+ * Interface to support listening for data received by the cm11a. This needs to be sent to OpenHAB so it has the current
+ * status of the modules
+ *
+ * @author Bob Raker
```
add "- Initial contribution"

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/internal/InvalidAddressException.java:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Exception for invalid X10 House / Unit code
+ *
+ * @author Bob Raker
```
add "- Initial contribution"

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/internal/X10Interface.java:
```diff
@@ -0,0 +1,804 @@
+ * Driver for the CM11 X10 interface.
+ *
+ *
+ * @author anthony green - Original code
```
Ah, here we go. Might you want to write his name capitalized?
As this is code from him, it would be good to get his approval [for our DCO](https://github.com/openhab/openhab2-addons/blob/master/CONTRIBUTING.md#sign-your-work) and you must list him with an "Also-By" line in your commit comment (right above the sign off).

BobRak replied:
Anthony Green wrote a whole openHAB-1 binding. I used this X10Interface with changes. The rest of it I started from scratch as an openHAB-2 binding. I make have copied lines from his binding. It has been over a year ago I started on this and I don't remember all of the details. As I indicated above I have been unable to contact him. I can add the Also-By line but getting his approval doesn't seem possible to me.

kaikreuzer replied:
No need to add any details on what you changed - it will be enough if you put something like "Bob Raker - adapted code for use in openHAB 2" or similar.

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/internal/X10Interface.java:
```diff
@@ -0,0 +1,804 @@
+ *
+ *
+ * @author anthony green - Original code
+ * @author Bob Raker-
```
add information about what you did to the class

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/internal/X10ReceivedData.java:
```diff
@@ -0,0 +1,151 @@
+ * Container for data received from the powerline by the cm11a interface. When data is received one of more of these
+ * objects is created and then passed to interested objects.
+ *
+ * @author Bob Raker
```
add "- Initial contribution"

BobRak commented on addons/binding/org.openhab.binding.cm11a/README.md:
```diff
@@ -0,0 +1,101 @@
+
+##Credit
+
+This binding is made possible because of the previous work done by "Anthony Green". His work can be seen at: <https://github.com/engineergreen/openhab/tree/cm11a-binding/bundles/binding/org.openhab.binding.cm11a>
```
I am unable to get any information about Anthony Green. If you look at his repo there is no way to open an issue or contact him. If you put: https://api.github.com/users/engineergreen/events/public in a browser it comes back with []. I would very much like to contact him to thank him and ask his permission. If you have any suggestions on how to contact him I would appreciate your guidance.

kaikreuzer replied:
Ok, thanks for the background infos. Ok, I think as he contributed the code under EPL to openHAB 1, we can assume his approval here - his authorship is kept in the relevant classes, so this is fine.
For the reasons mentioned above, I would still say that this section here should be removed from the README.

BobRak commented:
Thanks for the feedback. I am up-to-my-neck in home repair projects at the moment and will work on this as I get time.

kaikreuzer commented:
@engineergreen Just in case that you are notified about this and read this: We'd like to thank you for your work on the openHAB1 binding and just like to let you know that it has now made it into openHAB 2 as well - assuming your approval of [the DCO](https://github.com/openhab/openhab2-addons/blob/master/CONTRIBUTING.md#sign-your-work) with all due respect!

engineergreen commented:
Hello, That's awesome @BobRak I shall try it out.  @kaikreuzer  - yes I approve the DCO.

kaikreuzer commented:
Excellent, thanks for your feedback, @engineergreen!

BobRak commented:
I have addressed all of your comments. About the format of the README.md file. I made many updates to it based on the README files for other bindings. If you want more changes to this please provide a binding that I might use as a template.

kaikreuzer commented on addons/binding/org.openhab.binding.cm11a/build.properties:
```diff
@@ -0,0 +1,7 @@
+               .,\
+               OSGI-INF/,\
+               ESH-INF/,\
+			   about.html
```
can you replace tabs by spaces?

kaikreuzer commented:
@BobRak Please NEVER merge the master into your branch - instead, rebase your branch on the latest master; that will make all git handling of your PR much easier.

BobRak commented:
@kaikreuzer I would appreciate some advise about the correct Git command. Sorrys, I know I should not need help with this but I continue to struggle with Git. Is the following correct:

> Assume my remotes are:
> origin  https://github.com/BobRak/openhab2-addons.git (fetch)
> origin  https://github.com/BobRak/openhab2-addons.git (push)
> upstream        https://github.com/openhab/openhab2-addons.git (fetch)
> upstream        https://github.com/openhab/openhab2-addons.git (push)

```
git checkout master
git fetch upstream
git rebase upstream/master
git checkout 1566-cm11a
git rebase upstream/master
# make changes / additions on my branch
git add .
git commit
# push my changes to my fork
git push origin 1566-cm11a

```

kaikreuzer commented:
Looks about right. But disclaimer: I hardly use git on cli, but most of the time GUIs...

BobRak commented:
Completed the changes requested.  And pushed changes. This time I applied upstream changes using rebase instead of merge.

kaikreuzer commented:
Many thanks again, @BobRak!
